mat64/dense.go

```diff
@@ -497,3 +500,71 @@ func (m *Dense) Augment(a, b Matrix) {
 	w := m.View(0, ac, br, bc).(*Dense)
 	w.Copy(b)
 }
+
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Dense is little-endian encoded as follows:
+//  0 -  8  number of rows    (int64)
+//  8 - 16  number of columns (int64)
+// 16 - ..  matrix data elements (float64)
+//          [0,0] [0,1] ... [0,ncols-1]
+//          [1,0] [1,1] ... [1,ncols-1]
+//          ...
+//          [nrows-1,0] ... [nrows-1,ncols-1]
+func (m Dense) MarshalBinary() ([]byte, error) {
+	var err error
+	buf := bytes.NewBuffer(make([]byte, 0, m.mat.Rows*m.mat.Cols*8+16))
+	err = binary.Write(buf, defaultEndian, int64(m.mat.Rows))
+	if err != nil {
+		return nil, err
+	}
+
+	err = binary.Write(buf, defaultEndian, int64(m.mat.Cols))
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < m.mat.Rows; i++ {
+		for _, v := range m.rowView(i) {
+			err = binary.Write(buf, defaultEndian, v)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return buf.Bytes(), err
+}
+
+// UnmarshalBinary decodes the binary form into the receiver.
+//
+// See MarshalBinary for the on-disk layout.
+func (m *Dense) UnmarshalBinary(data []byte) error {
+	buf := bytes.NewReader(data)
```

Comment:
I'm a little surprised we are being so unforgiving here, though it is consistent with the way we behave in other cases. @sbinet, we are essentially proxying type through the dims and allowing the zero-dims state be dynamic. Where it is surprising is that this is not how it is done pretty much anywhere else. Because of this, the behaviour should be explicitly documented in the doc comment.

For an (entertaining) explanation of why we are so brutal, when you have 45min to spare, watch this: https://www.youtube.com/watch?v=lrm8iEMQZNw